pkg/agent/plugin/nodeattestor/k8s/k8s.go

```diff
@@ -0,0 +1,245 @@
+package k8s
+
+import (
+	"context"
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/plugin/k8s"
+	"github.com/spiffe/spire/pkg/common/plugin/x509pop"
+	"github.com/spiffe/spire/proto/agent/nodeattestor"
+	"github.com/spiffe/spire/proto/common"
+	"github.com/spiffe/spire/proto/common/plugin"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/certificate/csr"
+)
+
+const (
+	pluginName = "k8s"
+)
+
+type configData struct {
+	spiffeID        string
+	privateKey      crypto.PrivateKey
+	attestationData *common.AttestationData
+}
+
+type K8sConfig struct {
+	TrustDomain     string `hcl:"trust_domain"`
+	PrivateKeyPath  string `hcl:"k8s_private_key_path"`
+	CertificatePath string `hcl:"k8s_certificate_path"`
+	K8sCACertPath   string `hcl:"k8s_ca_certificate_path"`
+	KubeconfigPath  string `hcl:"kubeconfig_path"`
+}
+
+type K8sPlugin struct {
+	m          sync.Mutex
+	c          *K8sConfig
+	kubeClient kubernetes.Interface
+}
+
+var _ nodeattestor.Plugin = (*K8sPlugin)(nil)
+
+func (p *K8sPlugin) getConfig() *K8sConfig {
+	p.m.Lock()
+	defer p.m.Unlock()
+	return p.c
+}
+
+func (p *K8sPlugin) setConfig(c *K8sConfig) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.c = c
+}
+
+func getAgentName() string {
+	name, err := os.Hostname()
+	if err != nil {
+		name = "unknown"
```

Comment:
There is an assumption in SPIRE that all agents can be uniquely addressed... as such, I think it would be best to raise an error here instead of set unknown since it seems that this name is ultimately used in the agent's spiffe id? Also brings up questions about hosts that might have the same hostname? Curious to hear your thoughts on this in the context of kubernetes cluster

More info here https://github.com/spiffe/spire/issues/558

Reply:
In a Kubernetes environment we can rely on each node (and pod) having a unique IP address because essentially NAT is prohibited, so if the trust domain coincides with the Kubernetes cluster, we could use the IP address.

https://kubernetes.io/docs/concepts/cluster-administration/networking/

If the trust domain includes multiple Kubernetes clusters, then I think the best we can do is to include a fingerprint of the CA certificate associated to the kubernetes cluster. 
Something like "spire/agent/k8s/cluster/XXXX/system:node:<host name>".

As a last resort, we can completely forget about the hostname and use a fingerprint of the cert, same as the x509pop attestation does.

I'm not sure what's the current thinking is on the question you referred.
Agent naming uniqueness is a nice property...
Human-readable SVIDs are nice too...

I will be happy to implement whatever you guys believe is the best overall solution.

Reply:
We discussed this topic on the SIG-Kubernetes call today. We concluded that it is probably OK for the resulting ID to be less human friendly than it could be. It is likely that a SPIRE server will manage multiple kubernetes clusters, so including the thumbprint of the CA seems like a good idea. The proposed path including thumbprint and node name is 👍.

Reply:
Server side node attestors can optionally kick back selectors which can be used to group agents together. In this case, it would be awesome if we provide a selector for CA thumbprint `k8s:ca_thumbprint:XXXXX`. Doing so would allow users to create SPIFFE IDs which address entire clusters, allowing them to register workloads to run in one cluster or another.